modules/table/src/main/java/org/apache/ignite/internal/table/distributed/LowWatermarkImpl.java

```diff
@@ -0,0 +1,280 @@
+    }
+
+    /**
+     * Schedule watermark updates.
+     */
+    public void scheduleUpdates() {
+        inBusyLock(busyLock, () -> {
+            HybridTimestamp lowWatermarkCandidate = lowWatermark;
+
+            if (lowWatermarkCandidate == null) {
+                LOG.info("Previous value of the low watermark was not found, will schedule to update it");
+
+                scheduleUpdateLowWatermarkBusy();
+
+                return;
+            }
+
+            LOG.info("Low watermark has been scheduled to be updated: {}", lowWatermarkCandidate);
+
+            txManager.updateLowWatermark(lowWatermarkCandidate)
+                    .thenComposeAsync(unused -> inBusyLock(busyLock, () -> notifyListeners(lowWatermarkCandidate)), scheduledThreadPool)
+                    .whenComplete((unused, throwable) -> {
+                        if (throwable == null) {
+                            inBusyLock(busyLock, this::scheduleUpdateLowWatermarkBusy);
+                        } else if (!(throwable instanceof NodeStoppingException)) {
+                            LOG.error("Error during the Watermark manager start", throwable);
+
+                            failureProcessor.process(new FailureContext(CRITICAL_ERROR, throwable));
+
+                            inBusyLock(busyLock, this::scheduleUpdateLowWatermarkBusy);
+                        }
+                    });
+        });
+    }
+
+    private @Nullable HybridTimestamp readLowWatermarkFromVault() {
+        VaultEntry vaultEntry = vaultManager.get(LOW_WATERMARK_VAULT_KEY);
+
+        return vaultEntry == null ? null : ByteUtils.fromBytes(vaultEntry.value());
+    }
+
+    @Override
+    public void stop() {
+        if (!closeGuard.compareAndSet(false, true)) {
+            return;
+        }
+
+        busyLock.block();
+
+        ScheduledFuture<?> lastScheduledTaskFuture = this.lastScheduledTaskFuture.get();
+
+        if (lastScheduledTaskFuture != null) {
+            lastScheduledTaskFuture.cancel(true);
+        }
+
+        IgniteUtils.shutdownAndAwaitTermination(scheduledThreadPool, 10, TimeUnit.SECONDS);
+    }
+
+    @Override
+    public @Nullable HybridTimestamp getLowWatermark() {
+        return lowWatermark;
+    }
+
+    void updateLowWatermark() {
+        inBusyLock(busyLock, () -> {
+            HybridTimestamp lowWatermarkCandidate = createNewLowWatermarkCandidate();
+
+            // Wait until all the read-only transactions are finished before the new candidate, since no new RO transactions could be
+            // created, then we can safely promote the candidate as a new low watermark, store it in vault, and we can safely start cleaning
+            // up the stale/junk data in the tables.
+            txManager.updateLowWatermark(lowWatermarkCandidate)
+                    .thenComposeAsync(unused -> inBusyLock(busyLock, () -> {
+                        vaultManager.put(LOW_WATERMARK_VAULT_KEY, ByteUtils.toBytes(lowWatermarkCandidate));
```

Comment:
I think it's better to save the new low watermark only after listeners complete their futures.

Reply:
This code left "as is" after renaming the class.

I disagree.
If we save watermark after notifying listeners, then we may loose LWM update and notify some listeners.
Thus we may read stale LWM on restart and some components may have destroyed data "in future" (above LWM).

Reply:
Ok